File: compiler/parser/TokenStream.cpp
#include <iostream>
#include "TokenStream.h"

using namespace MaximParser;

TokenStream::TokenStream(const std::string &data)
        : data(data), peekBuffer(Token(Token::Type::END_OF_FILE, "", SourcePos(0, 0), SourcePos(0, 0))) {
    restart();
}

void TokenStream::restart() {
    dataBegin = data.begin();
    hasBuffer = false;
    currentLine = 0;
    currentColumn = 0;
}

Token TokenStream::next() {
    if (hasBuffer) {
        hasBuffer = false;
        return peekBuffer;
    }
    auto nextToken = processNext();
    return nextToken;
}

Token TokenStream::peek() {
    if (!hasBuffer) {
        hasBuffer = true;
        peekBuffer = processNext();
    }

    auto nextToken = peekBuffer;

    return peekBuffer;
}

Token TokenStream::processNext() {
    if (dataBegin >= data.end()) {
        return Token(
                Token::Type::END_OF_FILE, "",
                SourcePos(currentLine, currentColumn),
                SourcePos(currentLine, currentColumn)
        );
    }

    bool hasToken = false;
    Token pickToken(Token::Type::UNKNOWN, "", SourcePos(0, 0), SourcePos(0, 0));

    do {
        for (const auto &token : TokenStream::matches) {
            std::match_results<std::string::iterator> m;
            if (!std::regex_search(
                    dataBegin, data.end(),
                    m, token.first,
                    std::regex_constants::match_continuous
            ))
                continue;

            dataBegin += m.length();

            auto tokenLine = currentLine;
            auto tokenColumn = currentColumn;

            if (token.second == Token::Type::END_OF_LINE) {
                currentLine++;
                currentColumn = 0;
            } else {
                currentColumn += m.length();
            }

            auto tokenContent = m.size() > 1 ? m[1].str() : "";
            hasToken = true;
            pickToken = Token(
                    token.second, tokenContent,
                    SourcePos(tokenLine, tokenColumn),
                    SourcePos(tokenLine, tokenColumn + m.length())
            );
            break;
        }
    } while (!filter(pickToken));

    if (hasToken) return pickToken;

    auto remainingStr = std::string(dataBegin, data.end());
    dataBegin = data.end();
    return Token(
            Token::Type::UNKNOWN, remainingStr,
            SourcePos(currentLine, currentColumn),
            SourcePos(currentLine, currentColumn)
    );
}

bool TokenStream::filter(const Token &token) {
    if (token.type == Token::Type::HASH) {
        isSingleLineComment = true;
    } else if (token.type == Token::Type::END_OF_LINE) {
        isSingleLineComment = false;
    } else if (token.type == Token::Type::COMMENT_OPEN) {
        multiLineCommentCount++;
    }

    auto isValid = !isSingleLineComment && !multiLineCommentCount;

    if (token.type == Token::Type::COMMENT_CLOSE) {
        multiLineCommentCount--;
        if (multiLineCommentCount < 0) multiLineCommentCount = 0;
    }

    return isValid;
}

TokenStream::PairListType TokenStream::matches = {
        // multi-char tokens
        getToken(R"(\.\.\.)", Token::Type::ELLIPSIS),
        getToken(R"(==)", Token::Type::EQUAL_TO),
        getToken(R"(!=)", Token::Type::NOT_EQUAL_TO),
        getToken(R"(<=)", Token::Type::LTE),
        getToken(R"(>=)", Token::Type::GTE),
        getToken(R"(\+=)", Token::Type::PLUS_ASSIGN),
        getToken(R"(-=)", Token::Type::MINUS_ASSIGN),
        getToken(R"(\*=)", Token::Type::TIMES_ASSIGN),
        getToken(R"(\/=)", Token::Type::DIVIDE_ASSIGN),
        getToken(R"(%=)", Token::Type::MODULO_ASSIGN),
        getToken(R"(\^=)", Token::Type::POWER_ASSIGN),
        getToken(R"(->)", Token::Type::CAST),
        getToken(R"(\+\+)", Token::Type::INCREMENT),
        getToken(R"(--)", Token::Type::DECREMENT),
        getToken(R"(\^\^)", Token::Type::BITWISE_XOR),
        getToken(R"(&&)", Token::Type::LOGICAL_AND),
        getToken(R"(\|\|)", Token::Type::LOGICAL_OR),
        getToken(R"(\/\*)", Token::Type::COMMENT_OPEN),
        getToken(R"(\*\/)", Token::Type::COMMENT_CLOSE),

        // multi-char identifiers
        getToken(R"(\bnum\b)", Token::Type::NUM_KEYWORD),
        getToken(R"(\bmidi\b)", Token::Type::MIDI_KEYWORD),
        getToken(R"(\bpure\b)", Token::Type::PURE_KEYWORD),
        getToken(R"(\bconst\b)", Token::Type::CONST_KEYWORD),

        // free tokens
        getToken(R"('((?:\\'|(?:(?!').))*)')", Token::Type::SINGLE_STRING),
        getToken(R"(\"((?:\\\"|(?:(?!\").))*)\")", Token::Type::DOUBLE_STRING),
        getToken(R"(((?=\.\d|\d)(?:\d+)?(?:\.?\d*)(?:[eE][+-]?\d+)?))", Token::Type::NUMBER),
        getToken(R"(:([a-gA-G]#?[0-9]+))", Token::Type::NOTE),
        getToken(R"(([_a-zA-Z][_a-zA-Z0-9]*))", Token::Type::IDENTIFIER),

        // single-char tokens
        getToken(R"(\+)", Token::Type::PLUS),
        getToken(R"(-)", Token::Type::MINUS),
        getToken(R"(\*)", Token::Type::TIMES),
        getToken(R"(\/)", Token::Type::DIVIDE),
        getToken(R"(%)", Token::Type::MODULO),
        getToken(R"(\^)", Token::Type::POWER),
        getToken(R"(=)", Token::Type::ASSIGN),
        getToken(R"(!)", Token::Type::NOT),
        getToken(R"(\()", Token::Type::OPEN_BRACKET),
        getToken(R"(\))", Token::Type::CLOSE_BRACKET),
        getToken(R"(\[)", Token::Type::OPEN_SQUARE),
        getToken(R"(\])", Token::Type::CLOSE_SQUARE),
        getToken(R"(\{)", Token::Type::OPEN_CURLY),
        getToken(R"(\})", Token::Type::CLOSE_CURLY),
        getToken(R"(,)", Token::Type::COMMA),
        getToken(R"(;)", Token::Type::SEMICOLON),
        getToken(R"(\.)", Token::Type::DOT),
        getToken(R"(:)", Token::Type::COLON),
        getToken(R"(#)", Token::Type::HASH),
        getToken(R"(>)", Token::Type::GT),
        getToken(R"(<)", Token::Type::LT),
        getToken(R"(&)", Token::Type::BITWISE_AND),
        getToken(R"(\|)", Token::Type::BITWISE_OR),

        getToken(R"(\n)", Token::Type::END_OF_LINE)
};

TokenStream::PairType TokenStream::getToken(const std::string &regex, Token::Type type) {
    return std::make_pair(
            std::regex(R"([^\S\n]*)" + regex + R"([^\S\n]*)", std::regex::ECMAScript | std::regex::optimize),
            type
    );
}

File: compiler/src/parser/token_stream.rs
use ast::{SourcePos, SourceRange};
use parser::{Token, TokenType};
use regex::Regex;
use std::iter::Peekable;

type TokenMatcher = (Regex, TokenType);

lazy_static! {
    static ref PAIR_MATCHES: [TokenMatcher; 48] = [
        // multi-char tokens
        get_matcher(r"\.\.\.", TokenType::Ellipsis),
        get_matcher(r"==", TokenType::EqualTo),
        get_matcher(r"!=", TokenType::NotEqualTo),
        get_matcher(r"<=", TokenType::Lte),
        get_matcher(r">=", TokenType::Gte),
        get_matcher(r"\+=", TokenType::PlusAssign),
        get_matcher(r"-=", TokenType::MinusAssign),
        get_matcher(r"\*=", TokenType::TimesAssign),
        get_matcher(r"/=", TokenType::DivideAssign),
        get_matcher(r"%=", TokenType::ModuloAssign),
        get_matcher(r"\^=", TokenType::PowerAssign),
        get_matcher(r"->", TokenType::Cast),
        get_matcher(r"\+\+", TokenType::Increment),
        get_matcher(r"--", TokenType::Decrement),
        get_matcher(r"\^\^", TokenType::BitwiseXor),
        get_matcher(r"&&", TokenType::LogicalAnd),
        get_matcher(r"\|\|", TokenType::LogicalOr),
        get_matcher(r"/\*", TokenType::CommentOpen),
        get_matcher(r"\*/", TokenType::CommentClose),

        // free tokens
        get_matcher(r"'((?:[^'\\]|\\.)*)'", TokenType::SingleString),
        get_matcher(r#""((?:[^"\\]|\\.)*)""#, TokenType::DoubleString),
        get_matcher(r"([-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)", TokenType::Number),
        get_matcher(r":([a-gA-G]#?[0-9]+)", TokenType::Note),
        get_matcher(r"([_a-zA-Z][_a-zA-Z0-9]*(?:\[\])?)", TokenType::Identifier),

        // single-char tokens
        get_matcher(r"\+", TokenType::Plus),
        get_matcher(r"-", TokenType::Minus),
        get_matcher(r"\*", TokenType::Times),
        get_matcher(r"/", TokenType::Divide),
        get_matcher(r"%", TokenType::Modulo),
        get_matcher(r"\^", TokenType::Power),
        get_matcher(r"=", TokenType::Assign),
        get_matcher(r"!", TokenType::Not),
        get_matcher(r"\(", TokenType::OpenBracket),
        get_matcher(r"\)", TokenType::CloseBracket),
        get_matcher(r"\[", TokenType::OpenSquare),
        get_matcher(r"\]", TokenType::CloseSquare),
        get_matcher(r"\{", TokenType::OpenCurly),
        get_matcher(r"\}", TokenType::CloseCurly),
        get_matcher(r",", TokenType::Comma),
        get_matcher(r";", TokenType::Semicolon),
        get_matcher(r"\.", TokenType::Dot),
        get_matcher(r":", TokenType::Colon),
        get_matcher(r"#", TokenType::Hash),
        get_matcher(r">", TokenType::Gt),
        get_matcher(r"<", TokenType::Lt),
        get_matcher(r"&", TokenType::BitwiseAnd),
        get_matcher(r"\|", TokenType::BitwiseOr),

        get_matcher("\n", TokenType::EndOfLine)
    ];
}

fn get_matcher(regex: &str, token_type: TokenType) -> TokenMatcher {
    let regex_str = format!(r"^[^\S\n]*{}[^\S\n]*", regex);
    (Regex::new(&regex_str).unwrap(), token_type)
}

struct TokenIterator<'a> {
    data: &'a str,
    cursor: usize,
    current_pos: SourcePos,
}

impl<'a> TokenIterator<'a> {
    pub fn new(data: &'a str) -> TokenIterator<'a> {
        TokenIterator {
            data,
            cursor: 0,
            current_pos: SourcePos { line: 0, column: 0 },
        }
    }
}

impl<'a> Iterator for TokenIterator<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Token> {
        if self.cursor >= self.data.len() {
            return None;
        }

        let token_start = self.current_pos;
        let remaining_data = &self.data[self.cursor..];

        let matched_pair = PAIR_MATCHES
            .iter()
            .filter_map(|matcher| {
                let maybe_caps = matcher.0.captures(remaining_data);
                match maybe_caps {
                    Some(caps) => Some((matcher.1, caps)),
                    None => None,
                }
            })
            .next();

        let token = match matched_pair {
            Some((token_type, captures)) => {
                let capture_length = captures[0].len();
                let token_end = if token_type == TokenType::EndOfLine {
                    SourcePos {
                        line: self.current_pos.line + 1,
                        column: 0,
                    }
                } else {
                    SourcePos {
                        line: self.current_pos.line,
                        column: self.current_pos.column + capture_length,
                    }
                };
                let token_content = if captures.len() > 1 { &captures[1] } else { "" };

                self.cursor += capture_length;
                self.current_pos = token_end;
                Token::new(
                    SourceRange(token_start, token_end),
                    token_type,
                    token_content.to_owned(),
                )
            }
            None => {
                // move cursor to end so the next iteration returns None
                self.cursor = self.data.len();

                Token::new(
                    SourceRange(token_start, token_start),
                    TokenType::Unknown,
                    "".to_owned(),
                )
            },
        };
        println!("Emit {:?}", &token);
        Some(token)
    }
}

pub type TokenStream<'a> = Peekable<Box<Iterator<Item = Token> + 'a>>;

pub fn get_token_stream<'a>(data: &'a str) -> TokenStream<'a> {
    let mut in_single_comment = false;
    let mut multi_comment_depth = 0;

    let boxed: Box<Iterator<Item = Token>> =
        Box::new(TokenIterator::<'a>::new(data).filter(move |token| {
            match token.token_type {
                TokenType::Hash => in_single_comment = true,
                TokenType::EndOfLine => in_single_comment = false,
                TokenType::CommentOpen => multi_comment_depth += 1,
                _ => {}
            }

            let is_valid = !in_single_comment && multi_comment_depth == 0;

            if token.token_type == TokenType::CommentClose && multi_comment_depth > 0 {
                multi_comment_depth -= 1;
            }

            is_valid
        }));

    boxed.peekable()
}
